solr/core/src/java/org/apache/solr/handler/component/ParallelHttpShardHandler.java

```diff
@@ -41,57 +39,78 @@
 @NotThreadSafe
 public class ParallelHttpShardHandler extends HttpShardHandler {
 
+  @SuppressWarnings("unused")
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
   private final ExecutorService commExecutor;
 
+  /*
+   * Unlike the basic HttpShardHandler, this class allows us to exit submit before
+   * pending is incremented and the responseFutureMap is updated. If the runnables that
```

Comment:
[Q] Taking the "this class allows us to exit submit before pending is incremented" behavior as a given, the use of these additional AtomicInt's makes sense.  And the descriptive comment here is lovely.

But why does the PR introduce this complication in the first place?  Is it required by SOLR-17158?  Is it fixing a concurrency bug that pre-exists SOLR-17158?

Reply:
The concurrency fixes here and changes in HttpShardHandler come from me needing to implement this and being worried about the effects of limits interrupting/canceling requests. I felt compelled to ensure that canceling futures from within the call back (to facilitate a faster short circuit once we know the request is hitting a limit) wouldn't cause concurrency issues. ... so I put effort into understanding exactly what's going on and what the controls are. I found that (before this ticket) there was reliance on the orderly update of pending, a map of futures and the request queue object. Converting the map to a synchronized object didn't seem sufficient since the logic relies on coordination among objects... so essentially I had to make sure everything was going to handle a new short circuit point initiated by the callback and concurrency became "on-topic"... 

At a high level this ticket works by letting the callbacks [signal to stop all other callbacks](https://github.com/apache/solr/pull/2666/files#diff-ce2580113ae9a9d50f308985229130f094868e83d35a8727c5683e6a2f3f44daR543) if one of them is failing and partial results are not desired. There's another opportunity to quit early in the take() method but that caused too many problems (rare test failures likely due to further introduced concurrency issues).

I do suspect some of this could be simplified, particularly what I note here: https://github.com/apache/solr/pull/2666/files#diff-ce2580113ae9a9d50f308985229130f094868e83d35a8727c5683e6a2f3f44daR98 

But I decided to only guard and de-duplicate the existing logic. I'm leaving revising the logic to a later effort in order to control scope. 

In the case of your class (which came in as a merge), this was the simplest thing I could think of (as in simple to understand, not simple/elegant). Basically we need to know that everything we scheduled has been attempted (fail or not) before we let the loop that takes responses exit or we might miss one, but we also have to ensure that a failure won't leave us looping forever. So the quick solution was to track what's started, and track what's been attempted, and make sure we don't quit waiting for responses unless they are equal.  Note that attempts may not result in a a future if `this.lbClient.requestAsync(lbReq);` fails, so the count of futures is not good enough.

Reply:
Ooof, the additional complexity I'm reading about is concerning.  Without looking at this as closely as either of you, admittedly, I've naively wondered why doesn't HttpShardHandler use ExecutorCompletionService.  It's a really nice JDK utility that seems to do conceptually what HttpShardHandler is doing.  If it would be helpful, it would reduce the complexity of what this Solr class has to deal with.  Thoughts?

> we need to know that everything we scheduled has been attempted (fail or not) before we let the loop that takes responses exit or we might miss one

That doesn't sound right.  Meaning, I hope we can cancel if we haven't actually sent all requests.

Reply:
That quote sounds scary because the code is complex and hard to describe. I am totally in agreement that we could refactor things further. (it was actually just refactored recently to use CompleteableFutures). However, I'm trying very hard to resist the (strong) temptation to completely redesign everything since I'm hoping a version of this can go in to 9x.

Here's a quick(?!?!) attempt to summarize (and possible first seed material of a blog post :) ):

The SearchHandler currently has 2 modes that it alternates until deciding it is done... First it enters a while loop that I will call [sending mode](https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java#L517C11-L552C1) where it sends every outgoing request it knows about using `shardHandler.submit(sreq, shard, params);`. Then it enters a while loop that I will call [receiving mode ](https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java#L557C1-L577C14) where it uses `ShardResponse srsp = tolerant ? shardHandler.takeCompletedIncludingErrors() : shardHandler.takeCompletedOrError()` to drain the responses that it has received unless something new shows up in outgoing. 

These two modes are contained in a while loop that is checking that there are no more outgoing requests. So the code reverts to send mode any time there's something to send, and exits if it finishes receiving with nothing further to send. This outer loop is needed because the receive process may queue new requests, but if we are done receiving no more requests can be queued, so at that point if nothing needs sending it's safe to continue.

In the classic single threaded HttpShardHandler, we can use a single atomic boolean incremented every time we send something and decremented when we receive something to keep track of how many things it was asked to send, and if we've received them all. 

In the new ParallelHttpShardHandler the send loop only schedules the send (to be performed by a thread in an executor before passing the thread, and the send mode loop will exit without any guarantee that even one http request is in flight yet. The Runnable that is passed to the executor calls `this.lbClient.requestAsync(lbReq);` and that method calls out into the actual http communication libraries which might be jetty http, jdk http, or whatever the latest fad for handling http is in the future. As such it doesn't seem wise to trust that it completes without exception. The callback is added to the future on the next line, but we can't be sure that that will happen.

Key take away: With parallel shard submit, we have two levels of tracking, one for initial intentions to make a request (at the SearchHandler level) and one for "has the Runnable been processed at all (at the shard handler level). 

This is important because [the `take(boolean)` based methods](https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java#L310C1-L318C4) have a loop that depends on obtaining a response for every thing we attempted. If we don't **track** what we attempt and **track** that the *attempt was made* (regardless of success/fail) then we run the risk of exiting that loop **before** we have ***started*** any of our http requests (in the executor threads which could delay/pause any amount of time)

So the sentence you quote about "needing to know everything we scheduled has been attempted", is referring specifically the new ParallelHttpShardHandler class question of whether or not the Runnable has been attempted by the Executor. So it's about _request initiation_ (remeber it's calling run**Async**()!), and has nothing to do with request completion.

Not discussed for simplicity:
 - tests of responses received vs the shard count that go hay wire the pending request count never exceeds the shard count.
 - fake responses created to satisfy the shard count test when shards are down.
 - empty strings as shard urls that inflate the shard count when shards are down necessitating the fake response.

Reply:
> As such it doesn't seem wise to trust that it completes without exception. The callback is added to the future on the next line, but we can't be sure that that will happen.

I followed you up until about here: what callback/lines are we talking about?  What failure/concurrency scenario are you envisioning with that?

I might be misunderstanding you entirely - but as I understand what you've said - that'd be the crux of the matter since it's what lead you to make the "pending" increment async, which in turn necessitates some of the other concurrency controls, etc.

Reply:
https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/ParallelHttpShardHandler.java#L91C11-L93C88

We schedule the async future with requestAsync, and then we add the whenComplete callback. This code should not trust the lbHttpClient method to return gracefully. An exception would leave us with no whenComplete() call, and then, the take loop would wait forever for a request that will never get added to requests.

Reply:
This code works, and it doesn't seem to have an impact on performance. If you can think of a more elegant solution, perhaps that can be a follow on ticket?

Reply:
> We schedule the async future with requestAsync, and then we add the whenComplete callback. This code should not trust the lbHttpClient.requestAsync method to return gracefully.

I see the bug or gap you're highlighting, but aren't there ways to handle that case without all of the additional counters?

e.g. Wrap the [lbHttpClient.requestAsync](https://github.com/apache/solr/blob/b748207bc331b5eeae284ee7602626dbe5e3ff50/solr/core/src/java/org/apache/solr/handler/component/ParallelHttpShardHandler.java#L91) in a try-catch that decrements the `pending` counter if the request-sending failed.  Or if you prefer, give the CompletableFuture returned by [CompletableFuture.runAsync](https://github.com/apache/solr/blob/b748207bc331b5eeae284ee7602626dbe5e3ff50/solr/core/src/java/org/apache/solr/handler/component/ParallelHttpShardHandler.java#L114C5-L114C31) a hook to handle that exception case.

Even seeing the gap, all these counters just don't seem needed to me.

Reply:
I'm not sure I've been clear. My response was explaining the section you quoted, but that's not the part that actually motivates the counters...  The key case those two counters guard against is this:

1. send/recieve [loop starts](https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java#L514)
2. send mode [loop starts](https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java#L517)
3. submit shard req 1 to Executor (waits for thread avail)
4. submit shard req 2 to Executor (waits for thread avail)
5. submit shard req 3 to Executor (waits for thread avail)
6. send mode loop exits, nothing left to send
7. receive mode [loop starts](https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java#L557)
8. recieve mode calls take(boolean) (via whichever route)
9. the [loop in take()](https://github.com/apache/solr/blob/d9d2af471610c66d56b43981215afec8e41ff65b/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java#L322) sees pending == 0 (because no reqs have executed yet) <<< SEED of PROBLEM
10. take loop exits, take(boolean) returns null.
11. receive mode loop exits, (nothing received so nothing new has been added to outgoing)
12. (receive/send alternating loop sees no further outgoing requests exits) <<< THE PROBLEM MANIFESTS
13. Main thread starts attempting to handle the (non-existent) responses
14. Thread becomes available, and Executor finally executes shard req 1 

The results from the parallel execution will be added to responses after some or all of the response processing has completed leading to nondeterministic results or errors, or in the worst case, we'd possibly return zero results (pretty sure current code would error with NPE, but that could get "fixed" later on, possibly when this shows up in someone's logs).

If we want to ensure that `pending` and `responseFutureMap` updates are synchronized for atomicity and "happens before" relationships, we have to do them both in the callback that runs asynchronously.

Reply:
Gotta love a 14-point list; thanks for being thorough!

> the loop in take() sees pending == 0 (because no reqs have executed yet) <<< SEED of PROBLEM

This all makes sense as things are written, but doesn't really explain why a simpler solution isn't possible. 
 If we need other counters because ParallelHttpShardHandler.submit() doesn't increment `pending` before returning...well, couldn't we modify submit() to behave that way?

Have `submit()` (or more accurately, `makeShardRequest`) increment the 'pending' counter before returning.  The runnable we create in `PHSH.makeShardRequest` would also have to decrement the counter in the error cases you've highlighted elsewhere in this thread (e.g. if the LB client returns an error).  But that'd work, seemingly (and be less complex than having multiple counters)?  Or is there still a gap with that approach that I'm missing?

Reply:
It partly depends on what you want to simplify. One of the original difficulties I identified was the fact that the three key objects: responses responseFutureMap and pending had an intricate and delicate relationship that was very difficult to reason about in the presence of multi-threading, and possible cancellation as well as exceptions being thrown. There are questions of order of operations, as well as memory model happens before relationships to consider, and the more code that stands between these objects and their coordinated modification, the more complicated the logic. Concurrency is hard, so I'm trying to simplify THAT aspect as much as possible. 

My strategy was to create minimal synchronized blocks that group updates to these objects, any time the responseFutureMap is in play, and generally move unrelated code out from between interactions with these objects. 

So at a high level, your suggestion would exchange 2 counters (relatively easy to reason about) with a separation across threads between the update to pending and the update to responseFutureMap.

That separation introduces the possibility that pending might be updated but if the the intervening code fails, there is no guarantee that a response will ever be placed in responses (note that just decrementing pending isn't enough, because THREE objects have to stay coordinated). The thread could stall indefinitely under load, die with OOM, or hit a bug in our code or a library... 

If that happens, the original request thread will be stuck in the loop in take() forever.

In this PR's code, **pending will never be updated unless the future has been created, and a whenComplete callback added** _which implies that a response will always be added to the responseFutureMap regardless of what happens down stream_, because the response is something we have already created and passed into the callback. (I'm trusting the prior code that there's no possibility of rsp=null and exception=null simultaneously here). 

So... 

- if we remove the counters with no other change we can skip past our requests and fail or give incorrect response.
- if we remove the counters and move pending as you suggest we avoid skipping past, but might poll forever in take(boolean) (or further reverting to prior code with blocking responses.take(), we would wait forever).

All of this suggests that the best solution is, as David mentioned, to rework all of this to use CompletionService. But that's too big a change for this PR IMHO.

Reply:
bq.  as David mentioned, to rework all of this to use CompletionService. 

Thanks for finally responding to this point; I wasn't sure if my idea resonated with anyone or not.

bq. But that's too big a change for this PR IMHO

It could fundamentally change the code and render your approach obsolete.  I hope we haven't dug ourselves a bigger tech-debt hole here; I dunno.  It'll be scarier for whoever takes this on, if it happens.

Reply:
> It could fundamentally change the code and render your approach obsolete. I hope we haven't dug ourselves a bigger tech-debt hole here; I dunno. It'll be scarier for whoever takes this on, if it happens.

Well to that end I've tried to leave comments clarifying things it took me hours to tease out and understand. If I can find free time or even sponsorship for that task, I'd certainly be happy to tackle it (soon would be excellent since I have my head wrapped around the issues now), but it also seems like something that radical shouldn't get ported to 9x, whereas I hope to move portions of this not pertaining to the parallel handler down to 9x (unless that got ported to 9x... I haven't actually checked yet).